File: #26-50 (Ada, C++, FORTRAN)/C++/Euler34.cpp
/////////////////////////////////////////////////////////////////////
// Euler34 ~ Digit factorials
// 
// Problem Statement:
// 
// 145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.
// 
// Find the sum of all numbers which are equal to the sum of the
// factorial of their digits.
// 
// Note: As 1! = 1 and 2! = 2 are not sums they are not included.
//
/////////////////////////////////////////////////////////////////////

/////////////////////////////////////////////////////////////////////
// Algorithm Explanation:
// 
// Uses modulo to extract digits then calculates the factorial for
// each and summing together. If the sum equals the full number it
// is added to the total sum.
//
/////////////////////////////////////////////////////////////////////

#include <iostream>
#include <chrono>

using namespace std;

int main()
{
	auto start = std::chrono::high_resolution_clock::now();
	int solution = 0;
	
	int cur_sum, digit, f, j;
	for(int i = 3; i < 10000000; i++)
	{
		j = i;
		cur_sum = 0;
		while(1)
		{
			f = 1;
			digit = j % 10;
			for (int k = digit; k > 0; k--)
			{
				f = f * k;
			}
			cur_sum += f;
			
			if(j < 10)
			{
				break;
			}
			j /= 10;
		}
		
		if(cur_sum == i)
		{
			solution += cur_sum;
		}
	}

	auto done = std::chrono::high_resolution_clock::now();
	cout << "Solution: " << solution << endl << endl;
	cout << "Program completed in " << std::chrono::duration_cast<std::chrono::milliseconds>(done - start).count() << " milliseconds." << endl;
}

File: #26-50 (Ada, C++, FORTRAN)/FORTRAN/Euler34.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Euler34 ~ Digit factorials
! 
! Problem Statement:
! 
! 145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.
! 
! Find the sum of all numbers which are equal to the sum of the
! factorial of their digits.
! 
! Note: As 1! = 1 and 2! = 2 are not sums they are not included.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Algorithm Explanation:
! 
! Uses modulo to extract digits then calculates the factorial for
! each and summing together. If the sum equals the full number it
! is added to the total sum.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

program Euler34

	implicit none
	integer :: time_start, time_finish, clock_finish, solution = 0
	integer :: cur_sum, digit, f, i, j, k

	call SYSTEM_CLOCK(time_start)

	do i = 3, 10000000
		j = i
		cur_sum = 0
		do while(.TRUE.)
			f = 1
			digit = modulo(j,10)
			k = digit
			do while(k > 0)
				f = f * k
				k = k - 1
			end do
			cur_sum = cur_sum + f
			
			if (j < 10) then
				EXIT
			end if
			
			j = j/10
		end do
		
		if (cur_sum == i) then
			solution = solution + cur_sum
		end if
	end do

	call SYSTEM_CLOCK(time_finish)
	clock_finish = time_finish - time_start
	write (*,"(A10,I10,//,A21,I5,A14)") "Solution: ",solution,"Program completed in ",clock_finish," milliseconds."

end program Euler34
